### PR TITLE
sql: add ON UPDATE syntax

### PR DESCRIPTION
**Pull request by pawalt**
Previously, we did not have parsing implemented for ON UPDATE
expressions on columns. This PR adds parsing for ON UPDATE expressions.

Release note: None

### CODE REVIEW

**Review by otan — APPROVED** [review]

File: pkg/sql/parser/sql.y [hunk 1]
```diff
@@ -6708,6 +6725,10 @@ col_qualification_elem:
   {
     $$.val = &tree.ColumnDefault{Expr: $2.expr()}
   }
+| ON UPDATE b_expr
+  {
+    $$.val = &tree.ColumnOnUpdate{Expr: $3.expr()}
```
Comment (otan) [inline pkg/sql/parser/sql.y:6730]: are we dealing with constraint name? there's some code here that assumes we are.

**Review by otan — CHANGES_REQUESTED** [review]
oops, missed a comment.

**Comment by otan** [comment]
i know this is late, but this FK stuff has be wondering - what we should do when we have `col ON UPDATE x REFERENCES other_table(col) ON UPDATE SET NULL`, and `other_table_col(col)` gets updated.

**Comment by pawalt** [comment]
@otan My interpretation is that the `ON UPDATE x` would be applied in that case. I think this falls under the same logic as `ON UPDATE CASCADE`. Any modification to a column results in the `ON UPDATE` being applied.

File: pkg/sql/parser/sql.y (hunk 1 above)
Comment (pawalt) [reply pkg/sql/parser/sql.y:6730]: Yes; because of the way the column qualifications are parsed, both the anonymous syntax `ON UPDATE expr` and the named syntax `CONSTRAINT name ON UPDATE expr` are valid uses.

**Comment by otan** [comment]
I think we should only have one ON UPDATE in the clause at any time, and/or prohibit FK ON UPDATE and ON UPDATE in the same column qualification.

Happy for that to be a separate PR. However, I feel like that also has knock on effects for parsing in that if you can guarantee one ON UPDATE in a qualification you can eliminate the ambiguity of ON UPDATE for FKs or ON UPDATE as a trigger. You can say if a FK is in the clause, the ON UPDATE bit is for the FK and only allow DEFAULT AND NULL as the expr, otherwise its the ON UPDATE trigger. That way we can keep the ON UPDATE SET clause.

If we are deviating from the RFC, we should update it accordingly.

File: pkg/sql/parser/sql.y (hunk 1 above)
Comment (otan) [reply pkg/sql/parser/sql.y:6730]: You should add a test in the Parser for this

**Comment by pawalt** [comment]
The multiple ON UPDATE concern should be taken care of by step where I build the `ColumnTableDef`. I check if there are multiple ON UPDATEs and error out if there is one.

I agree about prohibiting a FK ON UPDATE and an ON UPDATE in the same clause, but I think using that restriction to make parsing easier would be tough. From the parser perspective, the ON UPDATE reference action would have to be lifted to a top-level qualification, making it harder to construct a FK node since the ON UPDATE sits at the same level as the FK node.

This change would also make it impossible to specify an ON UPDATE expression with a value of DEFAULT/NULL for foreign key columns. To be clear, I don't think that this is a compelling use for ON UPDATE; I just don't like the semantics changing based on another qualification.

I don't have strong opinions on ON UPDATE from a SQL standards perspective, but I think in this case it makes behavior more explicit, and it's not a significant departure from ON UPDATE SET.

Once this PR gets merged in, I'll update the RFC with whatever gets decided here.

File: pkg/sql/parser/sql.y [hunk 2]
```diff
@@ -7088,13 +7109,13 @@ reference_actions:
...
 reference_on_update:
-  ON UPDATE reference_action
+  ON_UPDATE_LA UPDATE reference_action
   {
     $$.val = $3.referenceAction()
   }
 
 reference_on_delete:
-  ON DELETE reference_action
+  ON_DELETE DELETE reference_action
```
Comment (otan) [inline pkg/sql/parser/sql.y:7118]: can you name this `ON_DELETE_LA`

File: docs/generated/sql/bnf/stmt_block.bnf [hunk 3]
```diff
@@ -3044,10 +3050,10 @@ family_name ::=
 	name
 
 reference_on_update ::=
-	'ON' 'UPDATE' reference_action
+	'ON_UPDATE' 'UPDATE' reference_action
```
Comment (otan) [inline docs/generated/sql/bnf/stmt_block.bnf:3053]: sigh, i rarely read generated files on reviews but maybe i should
sorry to spring this on you also, but do you know why this isn't ON_UPDATE_LA?

Comment (otan) [reply docs/generated/sql/bnf/stmt_block.bnf:3053]: hmm, i guess the generated code omits the `_LA` as the other ones don't seem to be mentioned here.
maybe to make this clearer, it should be `ON_FK_UPDATE`

Comment (otan) [reply docs/generated/sql/bnf/stmt_block.bnf:3053]: thinking about this more, this might actually have to be just `ON_LA` so that the bnfs we use in our railroad diagrams in our docs just show `ON`. or we change the generated code to make this nicer...

File: pkg/sql/parser/lexer.go [hunk 4]
```diff
@@ -131,6 +135,16 @@ func (l *lexer) Lex(lval *sqlSymType) int {
 			case ALL:
 				lval.id = USER_ALL
 			}
+		case ON:
+			switch nextID {
+			case DELETE:
+				lval.id = ON_DELETE
```
Comment (otan) [inline pkg/sql/parser/lexer.go:141]: can you keep this as `ON`, or does it shift-reduce conflict?

Comment (pawalt) [reply pkg/sql/parser/lexer.go:141]: This ends up in conflict. From my testing, I always have to disambiguate the usage of ON or else I get a conflict.

File: docs/generated/sql/bnf/stmt_block.bnf (hunk 3 above)
Comment (pawalt) [reply docs/generated/sql/bnf/stmt_block.bnf:3053]: Oh good call. I just have to disambiguate between FK and ON UPDATE, so ON_LA for the FK use cases works here.

File: pkg/sql/parser/sql.y (hunk 2 above)
Comment (pawalt) [reply pkg/sql/parser/sql.y:7118]: Changed to ON_LA

File: pkg/sql/parser/sql.y (hunk 1 above)
Comment (pawalt) [reply pkg/sql/parser/sql.y:6730]: Done

File: docs/generated/sql/bnf/stmt_block.bnf (hunk 3 above)
Comment (otan) [reply docs/generated/sql/bnf/stmt_block.bnf:3053]: you still need a `make buildshort` to update this

File: pkg/sql/parser/testdata/create_table [hunk 5]
```diff
@@ -590,6 +590,30 @@ CREATE TABLE a (b INT8 DEFAULT (1)) -- fully parenthesized
...
+----
+CREATE TABLE a (b INT8 ON UPDATE 1)
+CREATE TABLE a (b INT8 ON UPDATE (1)) -- fully parenthesized
+CREATE TABLE a (b INT8 ON UPDATE _) -- literals removed
+CREATE TABLE _ (_ INT8 ON UPDATE 1) -- identifiers removed
+
+error
+CREATE TABLE foo(b INT8 ON UPDATE 1 REFERENCES blah (blah) ON UPDATE SET NULL)
+----
+at or near ")": syntax error: ON UPDATE expression and ON UPDATE foreign key action both specified for column "b"
```
Comment (otan) [inline pkg/sql/parser/testdata/create_table:604]: we also need to do this for ALTER TABLE ADD FOREIGN KEY / SET ON UPDATE. but we can do that in a separate commit (please make an issue)

**Review by otan — APPROVED** [review]

File: pkg/sql/parser/testdata/create_table (hunk 5 above)
Comment (pawalt) [reply pkg/sql/parser/testdata/create_table:604]: Yep I'm planning on adding this in the next PR since you can't just do this at the parsing level. #69052

**Comment by pawalt** [comment]
bors r=otan

**Review by ajwerner — COMMENTED** [review]
Can I get confirmation that we don't yet actually allow these clauses to be specified anywhere yet? Do we have reasonable gating around this feature? Have we talked about:

1) Updating the `catalog.Column` interface?
2) What can be referenced in these expressions?
    * Are we properly tracking them for uses of things like enum types?
3) Validating the descriptors with these things?

I feel like I'm not at all sure about what's going into this feature and I know there's been an RFC but it's been busy. I think I need to catch up. What's the status of this project. Are the work items broken out somewhere in a tracking issue?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68803#-:-MhKpxMZ-FjilssnWws4:be2p35a)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @arulajmani, @otan, and @pawalt)

**Comment by pawalt** [comment]
@ajwerner These clauses can be specified, but I don't think we need gating; in the morning I'll put up a PR implementing the logic for ON UPDATE. For your questions:

1. Yes, `HasOnUpdate` and `GetOnUpdateExpr` functions will have to be added.
2. References in ON UPDATE expressions follow the same rules as DEFAULT statements i.e. volatile expressions are allowed, but referencing other columns is not.
3. I'm adding validation logic to ensure that an ON UPDATE expression is not specified on a column with a foreign key update action. This is because it's ambiguous whether to apply the ON UPDATE expr or apply the FK action when a FK update happens.

Regarding issues, I haven't done a good job of tracking this, but I've thrown these up: #69052, #69057, #69058.
